### ISSUE TITLE
How to create configurations where the value is a file (text, binaries, cert)

### ISSUE THREAD

**Issue by gitricko** [epinio/epinio#2316]
**Is your feature request related to a problem? Please describe.**
Currently, the `epinio configuration create` is suitable for keys where values are simple string. 

How about creating a configuration key where the content is files ?

**Describe the solution you'd like**
We should have 2 ways to create this.

**cli**
`epinio configuration create <name> mykey -f ./myfile ` or something similar. 

**UI**

![Screenshot 2023-05-19 at 4 19 13 PM](https://github.com/epinio/epinio/assets/18060558/9f51b1bb-4201-4a13-9797-8223656eabd9)

**Describe alternatives you've considered**
We try to have some shell magic but it only uploaded the first line in a file

![signal-2023-05-19-120408_002](https://github.com/epinio/epinio/assets/18060558/56512de6-2bc8-48d7-95a7-0abe6fa917ed)

**Additional context**
Is there a short workaround now ?

**Comment by andreas-kupries** [comment]
Looking at your attempted workaround with `--set signing-key=$(cat ...)`,
did you try to wrap everything into quotes, i.e.  `--set signing-key="$(cat ...)"` ?
Or even   `--set "signing-key=$(cat ...)"` ?

**Comment by lhsilveira-ilia** [comment]
Managed to workaround this issue by encoding the file content to Base64:
`epinio configuration create [name] signing-key $(cat keys/signing.key.pem | base64 -w 0)`

And used procfile buildpack to define the startup script that will decode it:
`cat /configurations/[name]/signing-key | base64 -d > [path/to]/signing.key.pem`

However, when the output value of encoding to Base64 ends with "=", the update command fails:
- epinio configuration update test --set key1="value=="
- epinio configuration update test --set key1="$(echo value==)"
![image](https://github.com/epinio/epinio/assets/105882807/5ab85322-4b7e-4d36-8d46-c36322a14e38)

**Comment by mmartin24** [comment]
@richard-cox , could you please share any considerations to add this feature in UI?,  perhaps there is something done already in Rancher?

**Comment by enrichman** [comment]
@andreas-kupries  This morning we were thinking about some workaround to handle this but I've realized that it can still be implemented as a flag if we are going to implement it in the same way as `kubectl` does.
The `--from-file` flag accepts a `key=value` pair, to get the key and the file, so we don't need to change anything regarding the arguments.

```
    --from-file=[]:
	Key files can be specified using their file path, in which case a default name will be given to them, or
	optionally with a name and file path, in which case the given name will be used.  Specifying a directory will
	iterate each named file in the directory that is a valid secret key.
```

### key (--from-file=key=value)

The `--from-file=my-sec=mysecret.json` flag will create a `my-sec` key with the content of the `mysecret.json` file

### no key (--from-file=value)

If `value` is a folder (`--from-file=myfolder`) the cli will iterate through the folder looking for files, and it will use the name of the files as keys, and the content as values.

Otherwise if the `value` is a regular file `--from-file=mysecret.json` flag will create a `mysecret.json` key with the content of the `mysecret.json` file.

@richard-cox from a UI perspective the two last approaches seem complicated to me. I'm ok starting only with the first solution, and maybe handling the other in a future release. What do you think?

**Comment by richard-cox** [comment]
The new implementation for this, with screenshots, is https://github.com/rancher/dashboard/pull/9071 `0- Current Version`. There's an upload button which populates the `Value` text field with the contents of the file.

The generic key/value component we use does already have an option to supply all values from file in the following format
```
aKey=aValue
bKey=bValue
```

So from the UI world a lot of the core requirements are covered. 

To expand on this, particularly running through a directory, would be a lot more involved

**Comment by andreas-kupries** [comment]
I like the `--from-file key=path` form, and `--from-file path` where the key defaults to end element of the path.
I would not do directory iteration.
We already have similar options, see `--env a=b` for an example.

**Comment by andreas-kupries** [comment]
Ok. The basic implementation of a `--from-file` option is trivial. And works nicely for large values.
It did need some additional code for `show` to nicely truncate such large values.
All client side.

Now, binary looks to be a different kettle of fish. That looks to require API and server-side support.
From a first test of the existing code base the data looks to get double-encoded at some point.
Example:
```
# example binary data to stuff into a configuration
% cat BINARY10 | od -c
0000000 357 244   A 341 263   Z   P 305   W   E
0000012

# create the configuration
% epinio configuration create binary --from-file BINARY10

# look at the secret
% kubectl get secrets -n workspace binary -o yaml

# look at the decoded base64 found in the secret
% echo 77+977+9Qe+/ve+/vVpQ77+9V0U= | base64 -d | od -c
0000000 357 277 275 357 277 275   A 357 277 275 357 277 275   Z   P 357
0000020 277 275   W   E
0000024
```

I suspect that the issue is us sending the binary data through JSON in the clear and that that breaks something server side, possibly the unmarshalling.

Currently thinking about ways to extend this API endpoint in backward compatible ways.

**Comment by andreas-kupries** [comment]
Confirmed that marshaling the binary data is the issue, and the bytes are replaced with the UTF replacement character.

**Comment by andreas-kupries** [comment]
@enrichman @richard-cox Thoughts about the proposals below ?

Below a short specification of two ideas for fixing the issue with binary data.
Both are backward compatible, i.e. work with old clients (cli, UI), except for binary data sent by such old clients.

Note: Testing for binary value is done with `unicode/utf8.ValidString()`.

### Global signaling

  1. The structure `ConfigurationCreateRequest` is  extended with a boolean flag `isBase64Encoded`. Default is `false`.
  2. When the client sees at least one binary value the flag is set (`true`) and all values are sent encoded.
  3. The server decodes the base64 if signaled.
  4. The same thing happens in return, for structure `ConfigurationShowResponse` and transmissions from server to client, with their roles in 2. and 3. exchanged.

Good: Minimal signaling.
Bad: Everything is encoded, even if only a single value is binary.

### Local signaling

  1. The client encodes binary values, and signals in-band (.i.e. in the string), by adding the prefix `base64://`.
     To handle non-binary values with this prefix such values are encoded too.
     No other values are encoded.
  2. The server decodes only those values which are signaled as encoded, per the prefix.
  3. The same thing happens in return, for `ConfigurationShowResponse`, and transmissions from server to client, with their roles in 1. and 2. exchanged.

Good: Encodes only values where this is needed, although the set of values is more than just binary data.
Bad: Signaling takes more space.

**Comment by enrichman** [comment]
I don't think that the client or the server should own the responsability of knowing and manipulating the data in such a way.
I would just support "plaintext" data, and if needed the data should be encoded before sending them. The application and the owner of the config should know what is inside the configuration value.

i.e. if I need to push a binary I will just encode it somehow in base64 before sending it:

```
base64 mybin > encoded-file.txt
epinio configuration create binary --from-file mybin=encoded-file.txt --from-file config=config.json
```

If this is really needed then we can think about adding a `--from-binary-file` flag, that will encode the file content, but still, I would not try to mess with ALL the values encoding. This is going to hurt us in the future, for sure.

Basically I would not add this feature.

Agree on skipping the folder iteration. Not really needed as well.

**Comment by andreas-kupries** [comment]
@gitricko Is @enrichman's proposal a workable solution for you ?

**Comment by andreas-kupries** [comment]
PR goes with @enrichman , no handling of binary data, just loading from files, and large values.
If binary truly becomes needed it can be added later.
